src/test/com/Elessar/app/client/MyClientIntegTest.java

```diff
@@ -0,0 +1,233 @@
+
+
+    @BeforeClass
+    public static void setUpBeforeClass() throws Exception {
+        final String serverAddress = "127.0.0.1";
+        final int serverPort = 9000;
+        final String serverURL = new StringBuilder().append("http://")
+                                                    .append(serverAddress).append(":")
+                                                    .append(serverPort).toString();
+        clientA_Port = 4000;
+        clientB_Port = 5000;
+        final MetricManager serverMetricManager = new MetricManager("ServerMetric", 1000000);
+        final MetricManager clientMetricManager = new MetricManager("ClientMetric", 1000000);
+
+        // Setup server
+        mongoDB = MongoClients.create("mongodb://localhost:27017").getDatabase("test");
+        mongoDB.getCollection(MyDatabase.USERS).createIndex(Indexes.text(User.NAME), new IndexOptions().unique(true));
+
+        final MyDatabase db = new MongoDB(mongoDB, serverMetricManager);
+        final MyServer server = new MyServer("localhost", serverPort, db, serverMetricManager);
+
+        // Setup client A
+        final BlockingQueue<String> msgQueueA = new LinkedBlockingQueue<>();
+        clientA = new MyClient(serverURL, clientMetricManager);
+        final MyClientServer clientA_Server = new MyClientServer("localhost", clientA_Port, msgQueueA, clientMetricManager);
+
+        // Setup client B
+        final BlockingQueue<String> msgQueueB = new LinkedBlockingQueue<>();
+        clientB = new MyClient(serverURL, clientMetricManager);
+        final MyClientServer clientB_Server = new MyClientServer("localhost", clientB_Port, msgQueueB, clientMetricManager);
+
+        server.run();
+        clientA_Server.run();
+        clientB_Server.run();
+    }
+
+
+    @Test
+    public void test() {
+        registerTest();
+
+        logOnTest();
+
+        logOffTest();
+
+        p2pMsgTest();
+    }
+
+    private void registerTest() {
+        try {
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(or(eq(User.NAME, "Shuai Hu"),
+                    eq(User.NAME, "Zhi Xu"))).iterator()) {
+                assertEquals(false, cursor.hasNext());
+            }
+
+            RegistrationResponse registerResp1 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
+            assertEquals(true, registerResp1.getSuccess());
+
+            RegistrationResponse registerResp2 = clientA.register("Shuai Hu", "4238902", "shuaih@yahoo.com", "62357492");
+            assertEquals(false, registerResp2.getSuccess());
+
+            RegistrationResponse registerResp3 = clientA.register("Zhi Xu", "213252", "zhix@126.com", "62258237");
+            assertEquals(true, registerResp3.getSuccess());
+
+            try (MongoCursor<Document> cursor = mongoDB.getCollection(MyDatabase.USERS).find(eq(User.NAME, "Shuai Hu")).iterator()) {
+                assertEquals(true, cursor.hasNext());
```

Comment:
Also check:
- total number of documents returned is 1
- password, email and phone number are the same as requested

Same for the other user